src/edu/washington/escience/myria/operator/CSVFileScanFragment.java

```diff
@@ -181,142 +187,184 @@ protected TupleBatch fetchNextReady() throws IOException, DbException {
-                  source.getInputStream(partitionEndByteRange + 1, movingEndByte);
-              partitionInputStream =
-                  new SequenceInputStream(beginningOfRecord, concatenateEndOfRecord);
-              parser =
-                  new CSVParser(
-                      new BufferedReader(new InputStreamReader(partitionInputStream)),
-                      CSVFormat.newFormat(delimiter).withQuote(quote).withEscape(escape),
-                      bytePositionAtBeginningOfRecord,
-                      0);
-              iterator = parser.iterator();
-              onLastRow = true;
-              record = iterator.next();
+
+      if (record != null) {
+        /*
+         * Here, we read the first byte of the stream if this is not the first worker. This is simply an optimization.
+         * If the previous character is not a new line, we discard the line.
+         */
+        if (lineNumber - 1 == 0 && partitionStartByteRange != 0) {
+          char currentChar = (char) partitionInputStream.read();
+          if (currentChar != '\n' && currentChar != '\r') {
+            discardedRecord = true;
+            /*
+             * This handles closes the stream in corner cases in which a worker might only have a fragment of a tuple as
+             * its entire partition. Since the server makes sure to provide workers a partition with a minimum size,
+             * this case could only happen with very large rows.
+             */
+            if (onLastRow) {
+              parser.close();
               break;
             }
-            dataChar = startStreamReader.read();
+          } else {
+            /*
+             * This handles the case in which the splitting occurs in the middle of \r\n.
+             */
+            currentChar = (char) partitionInputStream.read();
+            if (currentChar == '\n') {
+              discardedRecord = true;
+            } else {
+              partitionInputStream =
+                  source.getInputStream(adjustedStartByteRange, adjustedEndByteRange);
+            }
           }
-          startStreamReader.close();
         }
-      }
-      // This is a partial fragment, columns are missing and we need to read more bytes
-      if (record.size() < schema.numColumns()) {
-        if (lineNumber - 1 != 0 && !isLastWorker) {
-          onLastRow = true;
-          long bytePositionAtBeginningOfRecord = record.getCharacterPosition();
-          if (!fixingStartByte) {
-            fixingStartByte = true;
-            partitionStartByteRange += bytePositionAtBeginningOfRecord;
+
+        /*
+         * Here, if we are on the last row, we make sure to read the entire row until we either hit a new line or until
+         * we have read the entire file (again, this is for the case where a single worker might be reading a single
+         * large row that was split among other workers)
+         */
+        if (record.size() == schema.numColumns() && onLastRow && !isLastWorker) {
```

Comment:
There are two corner cases that I'm worried about:

1. The last field ends with a record separator right on the partition boundary. Not sure how we'd handle this except by doing as you suggest and reading the final byte directly. I think the simplest way to do that might be to use our old trick again and overlap the first additional range we fetch with the original partition by one byte, so the first byte of the next range would be the last byte of the partition. This should be infrequent enough that the overhead doesn't really matter.
2. The last field ends with a record separator inside the partition, but the first field of the next record is truncated on the partition boundary in such a way that it can't be parsed even as a partial record (see my example of a quoted string). Then we need to somehow know that this trailing fragment is really the "last record", even though `iterator.hasNext()` returns `false` after we parse the previous record. Again, this is just my assumption and it needs to be tested. The quoted string example should be quite straightforward to test.